### DISCUSSION TITLE
Best practice for logo usage with Shiny for Python

### DISCUSSION THREAD

**Discussion by thohan88** [posit-dev/brand-yml#61] [Q&A] [answered]
This is an awesome package, great work! 

I am wondering where to store and how to reference logos from `_brand.yml` in a shiny application. 

The documentation recommends storing logos adjacent to `_brand.yml`:

https://github.com/posit-dev/brand-yml/blob/58176bfa9f11e7ea2f4b718b8fe2f4425ee20f6c/docs/brand/logo.qmd#L19-L22 

My impression from the docs is that usage in an application would be something like this:

```python
from shiny import App, reactive, render, req, ui
from pathlib import Path
from brand_yml import Brand
theme = ui.Theme.from_brand(__file__)
brand = Brand.from_yaml(__file__)

app_ui = ui.page_navbar(
    ui.nav_panel(
        "Example",
            ui.input_slider("n", "N", 0, 100, 20),
            ui.output_text_verbatim("txt"),
    ),
    theme=theme,
    title=ui.img(src=str(brand.logo.small.path), height=35, class_="me-2")
)


def server(input, output, session):
    @render.text
    def txt():
        return f"n*2 is {input.n() * 2}"

app = App(app_ui, server)

```

But this would not work unless the path in `brand.log.small` is also mounted as a static directory (or in `/www` for express). If the directory needs to be mounted, then I guess the placement of logos does not really matter? 

I think most users would probably look at the [branded theming example] (https://shinylive.io/py/examples/#branded-theming), maybe this could be extended to include logo usage? 

Thanks again and apologies if I am misunderstanding the intended usage here.

**Comment by thohan88** [comment]
A related question:
I notice that the logos could contain entries for both dark and light theme. Is my assumption correct that these are just location placeholders and there is nothing in Shiny which picks these up and allow for automatic switching based on `ui.input_dark_mode()`?

**Reply by gadenbuie** [reply, marked as answer]
Thanks for the question @thohan88! You're right in that we haven't developed that part of the brand.yml integration with Shiny. But I've started working on it and have a draft PR together that you could try out if you want (https://github.com/posit-dev/brand-yml/pull/62).

In that PR, both the standard logo resource and the light/dark logo resource gets a `.to_html()` method that you can use to include the logo somewhere in your app. An example use will look something like this:

```python
from shiny import App, render, ui

theme = ui.Theme.from_brand(__file__)

app_ui = ui.page_navbar(
    ui.nav_panel(
        "Input Output Demo",
        ui.layout_sidebar(
            ui.sidebar(
                "Sidebar inputs here...",
                title = theme.brand.logo.to_html(which="medium") if theme.brand.logo else ""
            ),
	),
	title=ui.TagList(
        theme.brand.logo.to_html(which="small", height="32px") if theme.brand.logo else None,
        " brand.yml Demo"
    )
    theme = theme
)
```

Note that `ui.Theme.from_brand()` includes the parsed brand, which you could also get with `brand_yml.Brand.from_yaml(__file__)`.

From the `brand` object, you can call `brand.logo.to_html(which="small")` to get the small logo, which will take into account light/dark if used for that size. If `brand.logo.small` isn't defined, it will error. That approach is functionally equivalent to calling `brand.logo.small.to_html()`, but you'll likely have to do some guarding for your type checker to get down to `brand.logo.small`.

The syntax isn't set in stone yet and I'd be happy to hear what you think if you have a chance to take it for a spin!

**Comment by thohan88** [comment]
Thanks, @gadenbuie! 

That is a neat interface, and the PR worked exactly as intended for me. 

I'm wondering whether it's possible to avoid having this CSS generated inline? 

```css
<style>
[data-bs-theme="light"] [data-when-theme-id="e3c15f19"][data-when-theme="dark"],
.quarto-light [data-when-theme-id="e3c15f19"][data-when-theme="dark"],
[data-bs-theme="dark"] [data-when-theme-id="e3c15f19"][data-when-theme="light"],
.quarto-dark [data-when-theme-id="e3c15f19"][data-when-theme="light"] {
    display: none;
}
</style>
``` 

I am not a CSS expert, but it feels like this could have been two global rules like:
```css
[data-bs-theme=light] [data-when-theme="dark"] { display: none; }
[data-bs-theme=dark] [data-when-theme="light"] { display: none; }
```
I notice the `data-when-theme-id`-attribute, so there may also be a bigger picture intended with this that I am not aware of.

Base64-encoding the images is nice to ensure it works out of the box. Would it be possible to make use of the `static_assets`-directory through `<img src=...` as well? 

Thanks again for the great work!

**Reply by gadenbuie** [reply]
Thanks for trying out the feature! You posed some excellent questions that I'm sure will come during internal review as well 😄 

> I'm wondering whether it's possible to avoid having this CSS generated inline? ... I am not a CSS expert, but it feels like this could have been two global rules like: ...

That would certainly work for Shiny – where `data-bs-theme="dark"` signals dark mode – but we also need `.quarto-dark` for Quarto scenarios, at least until Quarto uses Bootstrap's newer color mode features.

But the real reason I'm not using global CSS is because neither of these approaches work with `@media (prefers-color-scheme: "dark")` in the sense that if the global CSS included rules that use media queries to choose the color scheme, those rules will conflict with the directly-signaled color mode. E.g.: if the OS is set to dark color mode but the page or app is set to light, the light image would be hidden by the media query rules and the dark image would be hidden by the local app setting.

Both `data-bs-theme` and `.quarto-{dark,light}` are framework-specific color mode signals, but the media query is native web and available in all browsers. Obviously, we're focusing our efforts on developing brand.yml for Quarto and Shiny users, but I want it to be useful anywhere. That means we need users to be able to opt into following `prefers-color-scheme` or to choose their own selectors. That of course then leads to the need for each call to `logo.small.to_html()` to work separately – it'd be very confusing if making one call to `.to_html()` affected the other uses of the method.

Anyway, I very much wanted this to work with a single set of CSS rules but ended up with the inline CSS out of necessity.



> Base64-encoding the images is nice to ensure it works out of the box. Would it be possible to make use of the `static_assets`-directory through `<img src=...` as well?

Great catch with this as well! I also wanted to use `static_assets`, but unfortunately the way that it works in Shiny for Python makes it available to app authors but not for third-party extension developers. In other words, you as the app author can certainly add something like `static_assets={"brand": brand.path.parent}`, but I can't register the static directory from the `.to_html()` method automatically for the user. (In other words, we don't have a Python equivalent to the `shiny::addResourcePath()` that exists in Shiny for R.)

I briefly tried using `htmltools.HTMLDependency()` but this isn't a great approach for including static assets because the dependency creator can't control the location of the hosted assets. (I may raise this a feature request.)

Ultimately, base64-encoding will work anywhere that HTML is used (i.e. out of the box in Quarto, Shiny, and any other framework that can handle HTML). There a few tradeoffs – it'd certainly be more efficient to host a single file, especially if you re-use the same logo in many places – but the transportability is worth it (for now, I'll still keep thinking about higher-level fixes for this issue).

**Comment by thohan88** [comment]
Thanks, @gadenbuie! As suspected, you are much further in the thought process and beyond my CSS and browser understanding and capabilities. 

I completely understand and support the ambition to make it work anywhere and the decisions seem like good trade offs. Advanced users preferring more control could always make their own wrapper based on the functionality provided by `brand-yml` and basic users will probably not care much about whether the CSS is inline or not.

E.g. one use case I have is a single monochromatic SVG logo where fills and strokes are set to `currentColor`, with the actual colors being styled by CSS rules based on `color` for light/dark theme. This requires an inline SVG without an img tag. I do not think it is reasonable to expect such cases to be covered by `to_html()`, but it's still easy to achieve with _brand.yml using something like:
```python
# app.py
title=ui.div(
  ui.HTML(open(str(theme.brand.logo.small.path), "r").read()),
  class_="navbar-logo"
)
```

```yaml
# _brand.yml
  shiny:
    theme:
      preset: shiny
      rules: |
        .navbar-logo { height: 25px; display: flex; }
        .navbar-logo svg { color: $primary; }
        [data-bs-theme=dark] .navbar-logo svg { color: $brand_seagreen; }  
```

The current interface is simple and allows for extensibility, I like it! 

Going a bit off-topic: The only addition to `brand-yml` I could think of would be brand icons. I have worked with a couple of clients who have their own icon sets in SVGs. Something like `theme.brand.icon.{icon_name}.to_html()` or similar would be brilliant.

**Reply by gadenbuie** [reply]
Great point about inlining SVGs -- that's also something I wanted to cover for the exact use-case you've mentioned. Maybe I could add a `method` argument with something like `method="inline"` to default to inlining SVGs or base64-encoding other images. `method="base64"` would always encode and `method="relative"` would use the relative path, allowing you to use static paths...



> The only addition to `brand-yml` I could think of would be brand icons. I have worked with a couple of clients who have their own icon sets in SVGs. Something like `theme.brand.icon.{icon_name}.to_html()` or similar would be brilliant.

We've talked about adding an `assets` top-level key, but in the mean time you can do exactly this with `logo.images`:

```yaml
logo:
  images:
    icon-bookmark-fill: icons/bookmark-fill.svg
    icon-calendar: icons/calendar.svg
```

```python
theme.brand.logo.images["icon-bookmark-fill"].to_html()

# or to inline the SVG
ui.HTML(
    theme.brand.logo.images["icon-calendar"].path.absolute().read_text()
)
```

That last example uses the fact that `logo.images` is a dictionary of `BrandLogoResource`s, each of which have a `.path` property with a `.relative()` and `.absolute()` method that return a `Path` to the relative or absolute file location, respectively (relative here means _relative to the `_brand.yml` file_), and the `Path` object has a convenient `.read_text()` method.

If we inline SVGs, we'll probably use something equivalent to this:

```python
ui.span(
    theme.brand.logo.images["icon-calendar"].path.absolute().read_text(),
    style="width: 32px; display: inline-block;"
)
```
